File: cebu-server/app/inference/InferenceService.java
package inference;

import java.util.Collection;
import java.util.Map;

import org.openplans.tools.tracking.impl.Observation;

import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;

import com.google.common.collect.LinkedHashMultimap;
import com.google.common.collect.Maps;
import com.google.common.collect.Multimap;

/**
 * This class is an Actor that responds to LocationRecord messages and processes.
 * Note: this is essentially a thread(instance)
 * 
 * @author bwillard
 * 
 */
public class InferenceService extends UntypedActor {

  private static final Map<String, InferenceInstance> vehicleToInstance = Maps
      .newConcurrentMap();

  private static final Multimap<String, InferenceResultRecord> vehicleToTraceResults = LinkedHashMultimap
      .create();

  private final LoggingAdapter log = Logging.getLogger(getContext().system(),
      this);

  public static void clearInferenceData() {
    vehicleToInstance.clear();
    vehicleToTraceResults.clear();
  }
  
  public static void processRecord(Observation observation) {

    final InferenceInstance ie = getInferenceInstance(observation
        .getVehicleId());

    ie.update(observation);

    final InferenceResultRecord infResult = InferenceResultRecord
        .createInferenceResultRecord(observation, ie);

    vehicleToTraceResults.put(observation.getVehicleId(), infResult);

  }
  
  @Override
  public void onReceive(Object location) throws Exception {
    synchronized (this) {
      if (location instanceof Observation) {
        final Observation observation = (Observation) location;
        processRecord(observation);
        
        log.info("Message received:  "
            + observation.getTimestamp().toString());
      }
    }

  }

  public static InferenceInstance getInferenceInstance(String vehicleId) {
    InferenceInstance ie = vehicleToInstance.get(vehicleId);

    if (ie == null) {
      ie = new InferenceInstance(vehicleId);
      vehicleToInstance.put(vehicleId, ie);
    }

    return ie;
  }

  public static Collection<InferenceResultRecord> getTraceResults(
    String vehicleId) {
    return vehicleToTraceResults.get(vehicleId);
  }

}

File: cebu-rails/lib/inference/inference_service.rb
require 'monitor'
require 'java'
java_import org.openplans.tools.tracking.impl.util.OtpGraph

class InferenceService
  include MonitorMixin

  @@graph = OtpGraph.new

  def self.graph
    return @@graph
  end

  def initialize
    super
    @vehicleToInstance = {}
    @vehicleToTraceResults = {}
  end
  
  @@instance = InferenceService.new
  def self.instance
    @@instance
  end

  def clearInferenceData
    @vehicleToInstance = {}
    @vehicleToTraceResults = {}
  end
  
  def processRecord(observation)
    ie = getInferenceInstance(observation.getVehicleId())
    ie.update(observation)
    infResult = InferenceResultRecord.createInferenceResultRecord(observation, ie)
    
    @vehicleToTraceResults[observation.getVehicleId()] ||= []
    if !@vehicleToTraceResults[observation.getVehicleId()].member? infResult
      @vehicleToTraceResults[observation.getVehicleId()] << infResult
    end
  end
  
  def processRecordFromWeb(observation)
    synchronize do
      processRecord(observation)
      Rails.logger.info("Message received:  " + observation.getTimestamp().to_s)
    end
  end
  
  def getInferenceInstance(vehicleId)
    ie = @vehicleToInstance[vehicleId]

    unless ie
      ie = InferenceInstance.new(vehicleId)
      @vehicleToInstance[vehicleId] = ie
    end

    return ie;
  end

  def getTraceResults(vehicleId)
    return @vehicleToTraceResults[vehicleId]
  end
  private_class_method :new
end
